subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+static const struct bt_data *ad_stream_current_ltv_get(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
+	const bool current_ltv_complete = (stream->current_ltv_offset == current_ltv->data_len + 2);
```

Comment:
I'm not sure I follow this either. How does comparing an offset to a data length determine if something is complete?

Reply:
If we have read out the entire LTV, it is complete. The data_len is the data length of the LTV. The offset is the amount of data read including type and length field (as documented)

Reply:
I'm sorry, but I don't understand that. 

```

	/* Current LTV offset contains the offset in the AD structure.
	 * The length and type are included in this offset.
	 */
```

`Current LTV offset contains the offset in the AD structure.` - What is the offset in this case? As is, what is it offsetting? 

Let's assume the simplest case, of a single LTV structure being set in the advertising data, where the value length (data_len) is 5.  
In that case, I'd expect `current_ltv_offset` to start at 0, but where clearly it is a complete LTV structure, but the comparison here would return false, as `0 != 5 + 2`. 

Can you please help me understand this better?

Reply:
For the case of your example, the advertising data is `[L, T, <5 bytes of data>]`. The `current_ltv_offset` indicates how much of this data that has been read. It starts out at 0, is incremented to 1 when the length field is read, and ends up at 7 when length, type, and value have been read.

`current_ltv_complete` here refers to if the reading of it is complete. Maybe it should be renamed to `done_reading_ltv` or something similar?

Reply:
Ahhh, Eureka moment :D 

So what we are doing here, is that we are incrementing the `ad_index` or the `data_index` in the beginning of the function if the previous call of the function finished an LTV. 

Since `current_ltv_offset` is updated outside the scope of this function, we can't do that increment at the end of this function. **But does also mean that in order to parse/read a single LTV structure, this function would need to be called twice?** 

Not sure what feedback @alwa-nordic gave regarding the refactoring of this, but since he approved it, this must have been what he wanted, although it does feel somewhat odd that we have a function `ad_stream_current_ltv_get` that depend on a caller `ad_stream_read` incrementing a specific struct field between calls. 

At this point, I think it was easier to read and understand before the refactoring, but with enough added comments and documentation for how these two (highly coupled) function need to work together, this would be OK.

Reply:
> But does also mean that in order to parse/read a single LTV structure, this function would need to be called twice

Yes you are right. 

The benefit is now that it can read more than one LTV at the time. We should there be able to reuse this function in other places as well given that we incorporate the shortening of advertising data name.